ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Extended.hs

```diff
@@ -53,21 +53,28 @@ applyExtLedgerState :: ( UpdateLedger blk
                        , ProtocolLedgerView blk
                        , HasCallStack
                        )
-                    => NodeConfig (BlockProtocol blk)
+                    => Bool -- ^ Was this block previously applied to the given ledger state?
+                    -> NodeConfig (BlockProtocol blk)
                     -> blk
                     -> ExtLedgerState blk
                     -> Except (ExtValidationError blk) (ExtLedgerState blk)
-applyExtLedgerState cfg blk ExtLedgerState{..} = do
+applyExtLedgerState prevApplied cfg blk ExtLedgerState{..} = do
```

Comment:
I'd be tempted to use a new data type for this rather than `Bool` and to do case analysis. It'd be harder to accidentally get the interpretation of the condition backwards if they're named constructors and we do explicit case analysis.

Reply:
Maybe `BlockValidationMode` is one you can reuse, or if the names are not right, make a new one here. Types are cheap.

Arguably `BlockValidationMode` is not right even though it'd be isomorphic. In this context it's about the pre-condition we want to know holds, whereas `BlockValidationMode` is what to do (as allowed by the pre-condition).

Reply:
Yeah I agree. I'll define a `newtype` for this. I won't reuse `BlockValidationMode` for the reason you mentioned and also because I don't think we should make use of a concrete `cardano-ledger` data type here.

Reply:
Added a type, `BlockPreviouslyApplied`.